Add Darknet Model

nicholaiTukanov opened the pull request:
- Currently uses dummy layer for layers that have not been created yet or still need more work
- Adds tiny yolov3 as the test data for parsing
- Edited UpSample2D to match new layer interface

nicholaiTukanov committed:
Add Darknet parser

- Currently uses dummy layer for layers that have not been created yet or still need more work
- Adds tiny yolov3 as the test data for parsing
- Editted UpSample2D to match new layer interface
- Ignore dev_darknet_parser branch since it has broken code

nicholaiTukanov commented:
Note that ARM build is currently failing. This is because of dev_new_layers has not been merged into dev yet. Also the CI will fail since the unit test for Upsample has not been fixed yet.

mcmillan03 commented:
Where did the cfg and weights files come from?
What was the license under which they were distributed?
We may want to consider writing our own cfg and weights files so we don't have to change our terms.

mcmillan03 commented on include/small/DarknetParser.hpp:
```diff
@@ -0,0 +1,511 @@
+    BufferT biases(block_info.c);
+
+    // has batch norm weights
+    if(block_info.bn) {
```
Need to discuss possibility of two bias vectors, and also BufferT's go out of scope and get deleted.

nicholaiTukanov replied:
Oh whoops, I will apply a fix for the scoping issue.

2 bias vectors? I haven't seen that before. Do you have a site that showcases that possibility?

nicholaiTukanov replied:
I haven't handled this case in my newest changes yet. This needs to be discussed further.

mcmillan03 commented on include/small/DarknetParser.hpp:
```diff
@@ -0,0 +1,511 @@
+// set up input shape of network
+template <typename BufferT>
+small::Layer<BufferT>* parse_network(small::shape_type &input, block_info_t block_info) {
+    input = {1UL, block_info.c, block_info.h, block_info.w};
```
parse batch field and make sure it is 1 or just ignore as done here?

nicholaiTukanov replied:
In my newest change, I did not add parsing for the batch field since it is only related to training.

mcmillan03 commented on include/small/DarknetParser.hpp:
```diff
@@ -0,0 +1,511 @@
+    weights.read(reinterpret_cast<char*>(&header[0]), 5*sizeof(int));
+    std::cout << "Header: " << header[0] << "." << header[1] << "." << header[2] << "\t" << (int64_t)header[3] << "\n";
+    BufferT weight_data(total_elems-5);
+    weights.read(reinterpret_cast<char*>(weight_data.data()), total_elems*sizeof(float));
```
This could overrun buffer by 5 floats?

nicholaiTukanov replied:
Fixed in newest changes.

mcmillan03 commented on include/small/DarknetParser.hpp:
```diff
@@ -0,0 +1,511 @@
+// function pointer array for parsing each block type
+// order must match enum
+template <typename BufferT>
+small::Layer<BufferT>*  (* parse_funcs [])(small::shape_type&, block_info_t) = {
```
Dont think this is necessary

mcmillan03 commented on include/small/DarknetParser.hpp:
```diff
@@ -0,0 +1,511 @@
+            break;
+        }
+
+        // parse the line
```
in v2.0 consider moving key/val parsing in side each parse_* function and get rid of block_info_t

mcmillan03 commented on include/small/DarknetParser.hpp:
```diff
@@ -0,0 +1,511 @@
+#include <small/UpSample2DLayer.hpp>
+
+#define PARSER_DEBUG 0
+
```
need namespace small.

mcmillan03 commented on include/small/DarknetParser.hpp:
```diff
@@ -0,0 +1,511 @@
+
+struct block_info_t {
+    // image size
+    uint32_t w=-1;
```
These are dangerous values. Consider 0 instead.

mcmillan03 commented on include/small/DarknetParser.hpp:
```diff
@@ -0,0 +1,511 @@
+    uint32_t w=-1;
+    uint32_t h=-1;
+    uint32_t c=-1;
+    uint32_t ic=-1;
```
what is ic used for?

nicholaiTukanov commented:
> Where did the cfg and weights files come from? What was the license under which they were distributed? We may want to consider writing our own cfg and weights files so we don't have to change our terms.

They came from this repo: https://github.com/pjreddie/darknet which has some interesting licensing but in general seems like it an MIT-like one.

mcmillan03 commented on include/small/DarknetParser.hpp:
```diff
@@ -0,0 +1,511 @@
+//****************************************************************************
```
Filename should either be the same as the class defined within, or the primary free function (exceptions exist).  In this case the file name and function name should be something like: parse_darknet_model.  The test filename would then be test_parse_darknet_model.cpp

nicholaiTukanov replied:
I have now implemented a Darknet model class and made the parser a private method.

nicholaiTukanov committed:
Fixed parsing for maxpooling to use F padding

nicholaiTukanov committed:
Added comment to described where cfg comes from

mcmillan03 reviewed (changes requested):
Consider removing enums, block_info and jump table and moving layer specific parsing into the corresponding parse_<layer>() functions.

Consider putting all of this code inside a DarkenetModel class (subclass of Model).

nicholaiTukanov committed:
Merge branch 'dev_new_layers' of github.com:CMU-SPEED/SMaLLFramework into dev_darknet_parser_2

nicholaiTukanov committed:
Merge branch 'dev' of github.com:CMU-SPEED/SMaLLFramework into dev_darknet_parser_2

nicholaiTukanov committed:
[WIP] Create Darknet Model class
- Rework darknet parser to be a method within the darknet class. The method builds a darknet model given config and weights files.
- Add a simple route layer to compute output shape.
- Remove parser test while working on redesign.

nicholaiTukanov committed:
Fix typos and bugs associated with typos

- Finish implementing parser
- Remove const from input shape in model class. Input shape for darknet will not be known till parse time.
- Add test for darknet. Currently only runs the parser.

nicholaiTukanov committed:
Remove old parsing code

nicholaiTukanov commented:
@mcmillan03 I implemented the changes that we discussed. I had remove the `const` from `input_shape` in the Model class since I needed to set it at parse time. Let me know what you think.

Important: there are some filter sizes in the darknet config that will cause the assumption of Co % Cob to fail. We will need to add cases for those.

nicholaiTukanov committed:
Merge branch 'dev' of github.com:CMU-SPEED/SMaLLFramework into dev_darknet_parser_2

nicholaiTukanov committed:
Make print prettier and housekeep

sei-rsteele commented:
> @mcmillan03 I implemented the changes that we discussed. I had remove the `const` from `input_shape` in the Model class since I needed to set it at parse time. Let me know what you think.
> 
> Important: there are some filter sizes in the darknet config that will cause the assumption of Co % Cob to fail. We will need to add cases for those.

Nicholai and I were wondering if you even need an input shape for the constructor in the first place, especially since you have a `create_model_and_buffers` method. I suppose this leads to more use cases in the future. However, it's not important for the demo.

nicholaiTukanov committed:
Remove performance tests

nicholaiTukanov committed:
Fix potential bugs when reading empty lines or comments and general housekeeping

mcmillan03 commented on include/small/Model.hpp:
```diff
@@ -70,7 +72,7 @@ class Model
 
 protected:
-    shape_type const             m_input_shape;
+    shape_type              m_input_shape;
```
Why was const removed?

nicholaiTukanov replied:
Input shape is not known until I parse the config file and the parse method is a part of the Darknet model. I removed `const` in order to be able to set the input shape in the parse method.

mcmillan03 commented on include/small/RouteLayer.hpp:
```diff
@@ -0,0 +1,72 @@
+        : Layer<BufferT>()
+    {
+#if defined(DEBUG_LAYERS)
+        std::cerr << "Add(batches:" << input1_shape[BATCH]
```
Update debug info to be specific to route: Add --> Route, and output information about both input shapes

nicholaiTukanov replied:
Fixed in [f419425](https://github.com/CMU-SPEED/SMaLLFramework/pull/7/commits/f4194253ec542140623dcae1d5440638c1b3ee99)

mcmillan03 commented on include/small/RouteLayer.hpp:
```diff
@@ -0,0 +1,72 @@
+#endif
+
+        // H and W must match
+        if (input1_shape[HEIGHT] != input2_shape[HEIGHT] ||
```
General question: what should be done about BATCH elements.

nicholaiTukanov replied:
Since we are only concating channel dimensions, we need to make BATCH, HEIGHT, and WIDTH are the same. I added a check to ensure BATCH is the same in both shapes.

mcmillan03 commented on include/small/RouteLayer.hpp:
```diff
@@ -0,0 +1,72 @@
+        std::vector<Tensor<BufferT> const *> input,
+        std::vector<Tensor<BufferT>*>        output) const
+    {
+        ///@todo Do nothing or throw? or concat here?
```
This should implement the concat or call a Concat function in interface_abstract.hpp that implements the functionality (even before there is abstract_layer support).  The latter should be done if there will eventually be microkernel support for this operation.

nicholaiTukanov replied:
This will be handled by another PR.

mcmillan03 commented on include/small/models/Darknet.hpp:
```diff
@@ -0,0 +1,530 @@
+        else if(inputs.size() == 2) { route = new RouteLayer<BufferT> (inputs[0], inputs[1]); }
+        else {
+            std::cerr << "ERROR: route layer must have 1 or 2 layers" << std::endl;
+            throw std::exception(); /// @todo: throw some exception
```
consider throwing std::format_error or std::invalid_argument

nicholaiTukanov replied:
Fixed in [f419425](https://github.com/CMU-SPEED/SMaLLFramework/pull/7/commits/f4194253ec542140623dcae1d5440638c1b3ee99)

mcmillan03 commented on test/test_darknet.cpp:
```diff
@@ -0,0 +1,42 @@
+    std::cout << "\n\nReading Darknet CFG from " << cfg << std::endl;
+    std::cout << "Reading Darknet weights from " << weights << std::endl;
+
+    small::Darknet<small::FloatBuffer> model(cfg, weights);
```
Put this within a try block. write one or more catch clauses for the various exceptions and reporting messages and TEST_ASSERT(false);

nicholaiTukanov replied:
I created the try block in [f419425](https://github.com/CMU-SPEED/SMaLLFramework/pull/7/commits/f4194253ec542140623dcae1d5440638c1b3ee99) but I only catch a general exception at the moment. I will leave this conversation open until I handle all the different possibilities.

mcmillan03 reviewed (changes requested):
Only a few minor suggestions.  If some take too long then be sure to add tasks in the next sprint.

nicholaiTukanov committed:
Add changes to address reviewer comments

nicholaiTukanov committed:
Add error checker to ensure cfg and weights file exist
